v4/export/dump.go

```diff
@@ -206,3 +237,59 @@ func dumpTable(ctx context.Context, db *sql.DB, database, table string) (TableDa
 		colTypes: colTypes,
 	}, nil
 }
+
+func buildSelectAllQuery(conf *Config, db *sql.DB, database, table string) (string, error) {
+	var query strings.Builder
+	query.WriteString("SELECT * FROM ")
+	query.WriteString(database)
+	query.WriteString(".")
+	query.WriteString(table)
+	if conf.SortByPk {
+		orderByClause, err := buildOrderByClause(conf, db, database, table)
+		if err != nil {
+			return "", err
+		}
+		if orderByClause != "" {
+			query.WriteString(" ")
+			query.WriteString(orderByClause)
+		}
+	}
+	return query.String(), nil
+}
+
+func buildOrderByClause(conf *Config, db *sql.DB, database, table string) (string, error) {
+	pkName, err := getPrimaryKeyName(db, database, table)
+	if err != nil {
+		return "", err
+	}
+	if pkName != "" {
+		// the table has primary key.
+		return fmt.Sprintf("ORDER BY %s", pkName), nil
+	}
+	switch conf.ServerInfo.ServerType {
+	case TiDBServerType:
+		return "ORDER BY _tidb_rowid", nil
+	default:
+		// ignore when there is no primary key
+		return "", nil
+	}
+}
```

Comment:
In TiDB, please always `ORDER BY _tidb_rowid` unless PKIsHandle. The default behavior without any ORDER BY clause should be sufficient. 

The purpose of "order by primary key" is to ensure the KV pairs generated are ordered, so overlapping data range will happen less. The ORDER BY clause shouldn't slow down the dumping speed in normal cases.

Reply:
Is there a convenient way to know whether a table is PKIsHandle?

As far as I know, "a table contains the primary key" cannot infer "PKIsHandle", because if the config option `alter-primary-key` is enabled, all new-created tables are PKIsNotHandle.
See: https://github.com/pingcap/tidb/pull/13229

Can we simply try to `select _tidb_rowid from t` to see if it returns an error?

Reply:
@tangenta 

> Can we simply try to `select _tidb_rowid from t` to see if it returns an error?

Yes, that's how mydumper checks if `_tidb_rowid` exists.

Or, if we trust TiDB to be always ordered by PK, just don't include the ORDER BY 🙃 (the checklist is most relevant when we dump single table in parallel).